boards: arm: add support for STM32F0DISCOVERY

nathantsoi opened the pull request:
Support the ST STM32F0DISCOVERY board with STM32F051R8T6 SoC

Tested with `samples/basic/blinky`

Includes flash config in `stm32f0_disco.rst` for use with mcuboot

Signed-off-by: Nathan Tsoi <nathan@vertile.com>

galak reviewed (changes requested):
Split this into 2 commits, one for the SoC and one for the board.

nathantsoi committed:
arch: arm: STM32F051X8 support

Add support for the STM32F051X8 processor on the STM32F0DISCOVERY board

Signed-off-by: Nathan Tsoi <nathan@vertile.com>

nathantsoi committed:
boards: arm: add support for STM32F0DISCOVERY

Support the ST STM32F0DISCOVERY board with STM32F051R8T6 SoC

Tested with:
  - `samples/basic/blinky`
  - `samples/basic/button`

Includes flash config in `stm32f0_disco.rst` for use with mcuboot

Signed-off-by: Nathan Tsoi <nathan@vertile.com>

erwango commented:
@nathantsoi : Please rebase vs latest master, it will allow to get rid of dts.fixup file.

dbkinder commented on boards/arm/stm32f0_disco/doc/stm32f0_disco.rst:
```diff
@@ -0,0 +1,141 @@
+.. _stm32f0_disco_board:
+
+ST STM32F0DISCOVERY
+################
```
heading underline needs to be at least as long as the title (add some more ###)

dbkinder commented on boards/arm/stm32f0_disco/doc/stm32f0_disco.rst:
```diff
@@ -0,0 +1,141 @@
+Overview
+********
+
+The STM32F0 Discovery development board with STM32F051R8T6 MCU
```
Complete sentences are preferred.  Let's combine these into:

    The STM32F0 Discovery development board uses an STM32F051R8T6 MCU and
    integrates the ST-LINK/V2-1 debugger and programmer.  It also comes with a
    comprehensive STM32 software HAL library and various packaged software
    examples.

dbkinder commented on boards/arm/stm32f0_disco/doc/stm32f0_disco.rst:
```diff
@@ -0,0 +1,141 @@
+- Board power supply: through USB bus or from an external 5 V supply voltage
+- External application power supply: 3 V and 5 V
+- Four LEDs:
+    - LD1 (red) for 3.3 V power on
```
need a blank like before this sub-list

dbkinder commented on boards/arm/stm32f0_disco/doc/stm32f0_disco.rst:
```diff
@@ -0,0 +1,141 @@
+********
+
+- STM32F051R8T6 microcontroller featuring 64 KB Flash memory, 8 KB RAM in an LQFP64 package
+- On-board ST-LINK/V2 with selection mode switch to use the kit as a standalone ST-LINK/V2 (with SWD connector for programming and debugging)
```
try to keep line length < 80 characters (unless you've got a long URL):

    - On-board ST-LINK/V2 with selection mode switch to use the kit as a standalone 
      ST-LINK/V2 (with SWD connector for programming and debugging)

dbkinder commented on boards/arm/stm32f0_disco/doc/stm32f0_disco.rst:
```diff
@@ -0,0 +1,141 @@
+
+Hardware
+********
+
```
Let's start with an introduction:

    The STM32 Discovery board features:

    - STM32F051R8T6 microcontroller ...

and make sure there's a blank line before the first list item.

nathantsoi replied:
not sure why github didn't pick up the diff, but note that this introduction has been added

dbkinder commented on boards/arm/stm32f0_disco/doc/stm32f0_disco.rst:
```diff
@@ -0,0 +1,141 @@
+    - LD4 (blue) for PC8 output
+- Two push buttons (user and reset)
+- Extension header for all LQFP64 I/Os for quick connection to prototyping board and easy probing
+- An additional board is provided which can be connected to the extension connector for even easier prototyping and probing.
```
break the long lines please

dbkinder commented on boards/arm/stm32f0_disco/doc/stm32f0_disco.rst:
```diff
@@ -0,0 +1,141 @@
+- An additional board is provided which can be connected to the extension connector for even easier prototyping and probing.
+- Comprehensive free software including a variety of examples, part of STM32CubeF0 package or STSW-STM32049 for legacy Standard Libraries usage
+
+More information about STM32F051R8 can be found here:
```
Better to write this as:

    More information about STM32F051R8 can be found in the `STM32F0x8 reference manual`_.

dbkinder commented on boards/arm/stm32f0_disco/doc/stm32f0_disco.rst:
```diff
@@ -0,0 +1,141 @@
+
+Default Zephyr Peripheral Mapping:
+----------------------------------
+- UART_1_TX : PA9
```
add a blank line before the list

dbkinder reviewed (changes requested):
Thanks for adding documentation. Here are a few suggested edits.

erwango commented on arch/arm/soc/st_stm32/stm32f0/soc.h:
```diff
@@ -55,6 +55,11 @@
 #endif
 
+/* For IMG_MANAGER */
+#if defined(CONFIG_SOC_FLASH_STM32)
```
#5919 Submitted to get it cleaned across whole tree

erwango commented on boards/arm/stm32f0_disco/Kconfig.board:
```diff
@@ -0,0 +1,11 @@
+# Kconfig - STM32F0DISCOVERY development board with STM32F051R8 MCU configuration
+#
+# Copyright (c) 2017 RnDity Sp. z o.o.
```
Remove this line

erwango commented on boards/arm/stm32f0_disco/Kconfig.defconfig:
```diff
@@ -0,0 +1,34 @@
+# Kconfig - STM32F0 development board with STM32F051R8T6 MCU
+#
+# Copyright (c) 2017 RnDity Sp. z o.o.
```
Remove this line

nathantsoi commented:
i think i have everyone's requested changes in. thanks all for the review and let me know if there is anything else

jmtorrespalma commented on dts/arm/st/stm32f051.dtsi:
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
```
This is inconsistent with other submitted files.
Should probably be:
 `Copyright (c) 2018 Nathan Tsoi`

erwango commented on boards/arm/stm32f0_disco/board.h:
```diff
@@ -0,0 +1,32 @@
+#define GREEN_LED_GPIO_PORT "GPIOC"
+#define GREEN_LED_GPIO_PIN  9
+
+/* LD2 blue LED */
```
LD4

erwango commented on boards/arm/stm32f0_disco/board.h:
```diff
@@ -0,0 +1,32 @@
+#define USER_PB_GPIO_PORT   "GPIOA"
+#define USER_PB_GPIO_PIN    0
+
+/* LD2 green LED */
```
LD3

erwango commented on boards/arm/stm32f0_disco/Kconfig.defconfig:
```diff
@@ -0,0 +1,33 @@
+# Test cases fail due to 'SRAM' region overflow.
+if GPIO
+
+config GPIO_STM32_PORTD
```
Thanks to move GPIO definition in arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.stm32f051x8.
Then disable in stm32f0_disco_defconfig

nathantsoi replied:
Thx for the feedback @erwango. I'm not sure I follow though. I think you're asking for something like:

```
diff --git a/boards/arm/stm32f0_disco/Kconfig.defconfig b/boards/arm/stm32f0_disco/Kconfig.defconfig
index 91ccdeda9..f2c417b3a 100644
--- a/boards/arm/stm32f0_disco/Kconfig.defconfig
+++ b/boards/arm/stm32f0_disco/Kconfig.defconfig
@@ -17,17 +17,4 @@ config UART_STM32_PORT_1

 endif # UART_CONSOLE

-# Disable ports D and F to allow kernel test cases common and
-# contex to run on stm32f0_disco.
-# Test cases fail due to 'SRAM' region overflow.
-if GPIO
-
-config GPIO_STM32_PORTD
-       default n
-
-config GPIO_STM32_PORTF
-       default n
-
-endif # GPIO
-
 endif # BOARD_STM32F0_DISCO
diff --git a/boards/arm/stm32f0_disco/stm32f0_disco_defconfig b/boards/arm/stm32f0_disco/stm32f0_disco_defconfig
index b64c9eb14..25dd94a10 100644
--- a/boards/arm/stm32f0_disco/stm32f0_disco_defconfig
+++ b/boards/arm/stm32f0_disco/stm32f0_disco_defconfig
@@ -30,6 +30,8 @@ CONFIG_PINMUX=y

 # GPIO Controller
 CONFIG_GPIO=y
+CONFIG_GPIO_STM32_PORTD=n
+CONFIG_GPIO_STM32_PORTF=n

 # Clock configuration
 CONFIG_CLOCK_CONTROL=y
```

Does that look close?

erwango replied:
@nathantsoi : Exactly

erwango commented on boards/arm/stm32f0_disco/Kconfig.defconfig:
```diff
@@ -0,0 +1,33 @@
+config GPIO_STM32_PORTD
+	default n
+
+config GPIO_STM32_PORTF
```
ditto

erwango reviewed (changes requested):
Some additional comments otherwise looks good

dbkinder reviewed (approved):
thanks!

galak commented:
@erwango look good to you?

erwango reviewed (approved):
Fine with me!
